File: bestAlbum.py
def solution(genres, plays):
    answer = []

    map = {}
    for i in range(len(genres)):
        if (genres[i] in map):
            map[genres[i]]['count'] += plays[i]
            map[genres[i]]['music'].append((i, plays[i]))
        else:
            map[genres[i]] = {'count' : plays[i], 'music': [(i, plays[i])]}
    genreRank = []
    for key in map:
        genreRank.append((key, map[key]['count']))
        map[key]['music'].sort(key=lambda music: music[1], reverse=True)
    genreRank = sorted(genreRank, key=lambda genre: genre[1], reverse=True)

    for genre in genreRank:
        for i in range(2):
            if(len(map[genre[0]]['music'])==1 and i==1):
                break
            answer.append(map[genre[0]]['music'][i][0])

    return answer

solution(['classic', 'pop', 'classic', 'classic', 'pop'], [500, 600, 150, 800, 2500])

File: programmers/javascript/bestAlbum.js
function solution(genres, plays) {
    let answer = [];

    let obj = {};
    let obj2 = {};
    for (let i=0; i<genres.length; i++) {
        let genre = genres[i];
        let play = plays[i];
        if (genre in obj) {
            obj[genre].push([play, i]);
            obj[genre].sort((a,b) => b[0]-a[0]);
        } else {
            obj[genre] = [[play, i]];
        }
        if (genre in obj2) {
            obj2[genre] += play;
        } else {
            obj2[genre] = play;
        }
    }
    let newGenres = [];
    for (let genre in obj2) {
        newGenres.push([obj2[genre], genre]);
    }
    newGenres.sort((a, b) => {
        return b[0]-a[0];
    });
    for (let arr of newGenres) {
        let genre = arr[1];
        for (let i=0; i<obj[genre].length && i<2; i++) {
            answer.push(obj[genre][i][1]);
        }
    }

    return answer;
}

console.log(solution(['classic', 'pop', 'classic', 'classic', 'pop'], [500, 600, 150, 800, 2500]));